library_generation/model/config_change.py

```diff
@@ -0,0 +1,155 @@
+        self.latest_value = latest_value
+        self.library_name = library_name
+
+
+class QualifiedCommit:
+    def __init__(self, commit: Commit, libraries: set[str]):
+        self.commit = commit
+        self.libraries = libraries
+
+
+class ConfigChange:
+    def __init__(
+        self,
+        change_to_libraries: dict[ChangeType, list[LibraryChange]],
+        baseline_config: GenerationConfig,
+        latest_config: GenerationConfig,
+    ):
+        self.change_to_libraries = change_to_libraries
+        self.baseline_config = baseline_config
+        self.latest_config = latest_config
+
+    def get_changed_libraries(self) -> Optional[list[str]]:
+        """
+        Returns library name of changed libraries.
+        None if there is a repository level change.
+        :return: library names of change libraries.
+        """
+        if ChangeType.REPO_LEVEL_CHANGE in self.change_to_libraries:
+            return None
+        library_names = []
+        for change_type, library_changes in self.change_to_libraries.items():
+            if change_type == ChangeType.GOOGLEAPIS_COMMIT:
+                library_names.extend(self.__get_library_name_from_qualified_commits())
+            else:
+                library_names.extend(
+                    [library_change.library_name for library_change in library_changes]
+                )
+        return library_names
+
+    def get_qualified_commits(
+        self,
+        repo_url: str = "https://github.com/googleapis/googleapis.git",
+    ) -> list[QualifiedCommit]:
+        """
+        Returns qualified commits from configuration change.
+
+        A qualified commit is a commit that changes at least one file (excluding
+        BUILD.bazel) within a versioned proto path in the given proto_paths.
+        :param repo_url: the repository contains the commit history.
+        :return: QualifiedCommit objects.
+        """
+        tmp_dir = "/tmp/repo"
+        shutil.rmtree(tmp_dir, ignore_errors=True)
+        os.mkdir(tmp_dir)
+        # we only need commit history, thus shadow clone is enough.
+        repo = Repo.clone_from(url=repo_url, to_path=tmp_dir, filter=["blob:none"])
+        commit = repo.commit(self.latest_config.googleapis_commitish)
+        proto_paths = get_proto_paths(self.latest_config)
+        qualified_commits = []
+        while str(commit.hexsha) != self.baseline_config.googleapis_commitish:
+            qualified_commit = ConfigChange.__create_qualified_commit(
+                proto_paths=proto_paths, commit=commit
+            )
+            if qualified_commit is not None:
+                qualified_commits.append(qualified_commit)
+            commit_parents = commit.parents
+            if len(commit_parents) == 0:
+                break
+            commit = commit_parents[0]
+        shutil.rmtree(tmp_dir, ignore_errors=True)
+        return qualified_commits
+
+    def __get_library_name_from_qualified_commits(self) -> list[str]:
```

Comment:
```suggestion
    def __get_library_names_from_qualified_commits(self) -> list[str]:
```

(This may imply more changes than just changing this line)

Reply:
Done.